File: stack/rain_water_trap.py

def calculate_trapped_water(heights: list[int]) -> int:  # LeetCode Q.42.
    # Height decreasing monotonic stack.
    stack: list[tuple[int, int]] = []  # Format: (height, idx).
    trapped_water = 0

    for idx, height in enumerate(heights):
        anchor_height = 0

        # My right side heights can't work with left side heights shorter than me.
        while stack and stack[-1][0] < height:
            partner_height, partner_idx = stack.pop(-1)
            trapped_water += (idx - 1 - partner_idx) * (partner_height - anchor_height)
            anchor_height = partner_height  # Anchors for a taller left height.

        if stack:  # I can always work with my previous non-shorter height (if exists).
            partner_idx = stack[-1][1]
            trapped_water += (idx - 1 - partner_idx) * (height - anchor_height)

            # My right side heights can't work with left side height as tall as me.
            if stack[-1][0] == height:
                stack.pop(-1)

        stack.append((height, idx))

    return trapped_water

File: stack/42_hard/rain_water_trap.cpp
#include <vector>
#include <stack>
using namespace std;

int calculateTrappedWater(vector<int> &heights) // LeetCode Q.42.
{
    stack<pair<int, int>> stack; // Format: (height, idx).
    int trappedWater = 0;

    for (int idx = 0; idx < heights.size(); idx++)
    {
        int anchor = 0;

        while (!stack.empty())
        {
            int height = min(heights[idx], stack.top().first) - anchor;
            int width = idx - 1 - stack.top().second;

            trappedWater += height * width;

            // Stop when seeing a remaining taller bar.
            if (stack.top().first > heights[idx])
                break;

            anchor = stack.top().first;
            stack.pop();
        }

        stack.push({heights[idx], idx});
    }

    return trappedWater;
}
